californium-core/src/main/java/org/eclipse/californium/core/network/KeyToken.java

```diff
@@ -0,0 +1,112 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.californium.core.network;
+
+import java.net.InetSocketAddress;
+
+import org.eclipse.californium.core.coap.Token;
+
+/**
+ * Implementation of client-local and not client-local tokens.
+ */
+public class KeyToken {
+
+	/**
+	 * Precalculated hash.
+	 */
+	private final int hash;
+	/**
+	 * Token.
+	 */
+	private final Token token;
+	/**
+	 * Peer's address.
+	 */
+	private final InetSocketAddress peer;
+
+	/**
+	 * Create key token.
+	 * 
+	 * @param peer peer's address. may be {@code null}, if key token is not
+	 *            client-local
+	 * @param token token
+	 */
+	public KeyToken(InetSocketAddress peer, Token token) {
+		this.token = token;
+		this.peer = peer;
+		int hash = token.hashCode();
+		if (peer != null) {
+			hash += peer.hashCode() * 31;
+		}
+		this.hash = hash;
+	}
+
+	/**
+	 * Get token.
+	 * 
+	 * @return token
+	 */
+	public Token getToken() {
+		return token;
+	}
+
+	/**
+	 * Get peer's address.
+	 * 
+	 * @return peer's address
+	 */
+	public InetSocketAddress getPeer() {
+		return peer;
+	}
+
+	@Override
+	public final int hashCode() {
+		return hash;
+	}
+
+	@Override
+	public final boolean equals(Object obj) {
```

Comment:
Usually to block unintended extensions. Some believe also in better performance.
I changed it to use Object instead of InetSocketAddress. That enables implementers to provide their own identity object and have their hash/equals there.

Reply:
Personnally I really don't like this idea to avoid users to extends methods. (I prefer a warning in the javadoc, because maybe there is some case we don't think where this could make sense)

(I'm a bit traumatized because I really suffer by the past when I was coding eclipse plugin and I was force to copy/paste several classes just because someone decide that I MUST not extends or access to this class or this fields :sweat_smile: )